### PR TITLE
Add a test to ensure reset does not collapse density matrix

### PR DESCRIPTION
**Pull request by daxfohl**
I was going to update ResetGate to be simpler and more reusable, by having act_on do a measuremnt and then applying a PlusGate corresponding to remaining qudit steps. This passed all existing tests. However I later realized that measuring a qubit could collapse entangled qubits as well, breaking density matrix representation. 

Here is a test that will break with the above implementation, to make sure nobody else tries the same.

### CODE REVIEW

**Comment by dabacon** [comment]
The test is correct, but I don't quite understand what this is protecting against.  The test is, 1) prepare |00>+|11>, and then reset on qubit 0, so this ends up in the density matrix 1/2(|00><00| + |01><01|).  It sounds like you tried to implement reset as measure and then conditional on the measurement apply an X (higher dimensional if needed).  I'm not sure how we can do this in Cirq right now because it has to be conditional the measurement result?

**Comment by daxfohl** [comment]
@dabacon Here is now Reset is currently implemented. You can see for the Clifford case it already does the reset and conditional X: https://github.com/quantumlib/Cirq/blob/2e3e5f392ec0e10204101f0c8747a21913e4608f/cirq-core/cirq/ops/common_channels.py#L723.

I saw that and though "well that's silly, the same logic applies to all simulators", so I tried removing the type check there and have that logic apply to all state types. There were qudit tests that failed, but once I fixed that up with a PlusGate, all remaining tests passed.

I almost created a PR with that change until I considered how the density matrix simulator was supposed to work and tried this test out. This test failed under that implementation. So I'm proposing adding this to the test suite to guard against this oversight.

**Review by dabacon — APPROVED** [review]
Approve just consider a slightly better name or comment on what the test is testing for.

File: cirq-core/cirq/sim/density_matrix_simulator_test.py [hunk 1]
```diff
@@ -527,6 +527,23 @@ def test_simulate_qudits(dtype: Type[np.number], split: bool):
     assert len(result.measurements) == 0
 
 
+@pytest.mark.parametrize('dtype', [np.complex64, np.complex128])
+@pytest.mark.parametrize('split', [True, False])
+def test_simulate_reset_does_not_collapse(dtype: Type[np.number], split: bool):
```
[the remark this answers is not in the document]
Comment (daxfohl) [reply cirq-core/cirq/sim/density_matrix_simulator_test.py]: "heralded" isn't used anywhere else in cirq, so I went with a long name `test_reset_one_qubit_does_not_affect_partial_trace_of_other_qubits`, which hopefully conveys the intention (and hopefully that intention is correct--I only learned about partial traces recently).

I tried changing the test to do an actual partial trace of a bell circuit result pre- and post- reset and compare them, and it worked, but just made the test harder to read IMO, so I reverted the test code to what it was originally.

**Pull request merged**